### ISSUE TITLE
Complexity in the design

### ISSUE THREAD

**Issue by norswap** [Kotlin/kotlin-coroutines#76]
I read through the proposal yesterday, and I'm happy to see that it is simpler than the first version.
It still feels too complex however. But the ideas are sound, and I think there is a much cleaner design struggling to get out. I think I can help with that.

I have in mind a design very similar (but not identical) to how async/await works in Javascript and Typescript. It would feature only three primitives: `async` (function/lambda keyword), `await`, and a `Promise`/`Future` class (I think only the usual `then` and `catch` method need to be exposed). It should handle all the use cases. I still have to go through the open issues however.

The advantages would be:
- a simpler model
- less reliance on libraries to implement the basic use cases
- familiarity for users of async/await in other languages
- ease of implementation for the javascript backend
- improved java interop

I have a bit of time next week and I'm willing to write a detailed proposal. Would it be welcome, or are some part of the current design already set in stone? Is it going to be evaluated without respects to sunk cost?

**Comment by elizarov** [comment]
Please, consider the case of an efficient `generate`/`yield` on JVM. The current version is indeed more complex than the version that uses only async/await/Promise as its primitive, but any kind of `Promise` is pretty heavy-weight, so it would neccessarily slow-down `generate`/`yield` pair. That is why mainstream languages that have `async`/`await` also have `generate`/`yield` and they _still_ don't support all the other use-cases (like `asyncGenerate`/'yield`) and require additional built-in primitives to handle them.

**Comment by norswap** [comment]
What is the cause of the inefficiency?

Here's what I have in mind, you can let me know if you consider it inefficient or not.
It requires laying out some background, however.

An `async` function is a function that returns a promise. So: `async fun foo(): String` desugars to `fun foo(): Promise<String>`, and the body of the function is rewritten as a state machine. Within that state machine, each statement of the form `await <EXPR>`, where the expression evaluates to a promise, is rewritten as `return <EXPR>.then { ... }`. The code within the braces re-enters the state machine at the proper location. AFAIU, this is the same as is explained in the [State Machine](https://github.com/Kotlin/kotlin-coroutines/blob/master/kotlin-coroutines-informal.md#state-machines) section.

Okay, generators now. Here is how I would implement `yield` and `generate`. Explanation will follow the code.

```Kotlin
class Yielder<T> (start: Yield<T>.() -> Promise<Unit>)
{
    var value: T? = null
    var proceed = Promise<Unit>()
    init { proceed.then { start() } }

    // NOTE: inline is necessary to be able to use await in function f
    inline fun yield (value: T) {
        this.value = value
        proceed = Promise<Unit>()
        await proceed;
    }
}

fun <T> generate (f: Yield<T>.() -> Promise<Unit>)
    = object: Sequence<T>
    {
        val yielder = Yielder<T>(f)

        // NOTE: alternate Sequence interface, but you get the idea
        // TODO: handle the termination of the sequence
        override fun next(): T {
            // NOTE: will compute next value synchronously
             yielder.proceed.resolve(Unit)
             return yielder.value as T
        }
    }
```

The idea is that we compute a value, then suspend on the `await` inside `yield()` (`yield()` has to be inline so that we know how to rewrite). When it needs another value, the sequence just re-starts the computation by resolving the promise that was generated when the previous value was computed.

Notice how we didn't need to annotate `f`, even though it is an async function: we just treat it as a regular function returning a promise.

So essentially, compute each new value requires:
- creating a promise beforehand
- calling a lambda (one that was generate by the rewrite, to re-enter the state machine)

Seems reasonable to me. We might even be able to reuse a promise by resetting it after it has been fulfilled.

Let me know what you think.

**Comment by elizarov** [comment]
A new promise on every `yield` invocation is extremely inefficient. The current implementation of `generate`/`yield` is optimised so that it does not create any objects on yield. This way, you can write your lazy collection transformations with `generate`/`yield` and get the same performance as using the standard Kotlin `Sequence` transformations.

Also, your design is C#/Dart-style. If you have `await` as a keyword, then you'll have to have `await for as a special case, etc, etc. That is a never-ending spiral. It does not compose with other language features.

**Comment by norswap** [comment]
> Also, your design is C#/Dart-style. If you have await as a keyword, then you'll have to have await for as a special case, etc, etc. That is a never-ending spiral. It does not compose naturally with other language features.

This is not a fair argument. You are always going to have primitives. I have outlined mines (async, await, and basic implementation of Promises), so my proposal has strictly less primitives than the current one. You can make `async` a function if that's the bother. I think it's helpful making it a keyword, because it *is* magical after some fashion: it is what causes the execution to suspend. But that's really not an important point (`async` being a function or a keyword)

> A new promise on every yield invocation is extremely inefficient.

You might have missed the section where I say:

> We might even be able to reuse a promise by resetting it after it has been fulfilled.

(And really, I was too cautious, as I don't see any reason we might not be able too.)

**Comment by elizarov** [comment]
There is no thread-safe way to reuse a promise object. You'll have to either give it GC or write your own, still expensive, thread-safe memory reclamation scheme. 

You'd also recommend to workout a use-case for asynchronous iteration. How would a for loop work without introducing yet another `await for` primitive?

**Comment by norswap** [comment]
> There is no thread-safe way to reuse a promise object. You'll have to either give it GC or write your own, still expensive, thread-safe memory reclamation scheme.

Indeed, but who says it should be thread-safe in this case?

If you have

```Kotlin
interface Promise<T> {
    fun <U> then (f: (T) -> U): Promise<U>
    fun catch (f: (Exception) -> T
    fun accept (value: T)
    fun reject (exception: Exception)
}
```

Then you can have thread-safe and non-thread-safe implementations. If usability (safe defaults) is an issue, you can even make `Promise` a thread-safe base class, and have `LocalPromise` sub-class which can be reset, a bit in the same way that `RuntimeException` extends `Exception`.

> I'd also recommend to workout a use-case for asynchronous iteration. How would a for loop work without introducing yet another await for primitive?

Maybe I'm missing something, but I thought this was being taken care of as part of the state machine transformation. It can certainly be done (resume the execution inside a loop). `await` can only be used inside `async` functions (or inside `inline` functions called from `async` functions.

**Comment by norswap** [comment]
Okay, I think I understand your concern for `for`, I suppose it's related to [asyncGenerate](https://github.com/Kotlin/kotlin-coroutines/blob/e94b8246e45894ee2331c067db1642dcf805b1c3/examples/asyncGenerate.kt).

My answer would simply be to build an `Iterable<Promise<T>>` in thise case, then simply use `await` within the loop body. It does entail creating an promise for each item iterated on, but in this case I think it's warranted.

It's not like a regular generator, which is just a more convenient/natural way to write an iterator, and so you expect the same performance characteristics. No, here you are waiting for something supplied by someone else. When you have this kind of coordination, I don't think an extra object instantiation is really significant.

**Comment by elizarov** [comment]
You are right about performance in async case. When you are dealing with asynchronous computations extra overhead is not important. 

However, I still don't see how can you ever make an efficient `generate`/`yield` on top of promises. There are at least three performance problems with promises: 

1. The promise is _conceptually_ non-reusable. You can only reuse a promise object if you are absolutely sure that nobody is holding any references to the `Promise` object, because the holder of the reference can invoke `then` at any time. Even in a single-threaded case figuring out when a Promise can be reused requires some GC-like approach.

2. The promise is _conceptually_ heavy-weight object. It must support multiple `then` and `catch` invocation. While they can be optimised for a single-waiter case, it will still have extra checks and adds conceptual complexity.

3. The promise is _conceptually_ asynchronous object. In JS all `then` and `catch` handlers are invoked asynchronously (i.e. later in event loop) for a good reason. Promise cannot invoke its `then` handler right await in the same stack frame, because that would lead to stack overflow on long-running loops.

Bottom-line: promise is too high-level and too domain-specific object to be part of the _general purpose_ language. It is Ok to be a part of _domain-specific_ language, but in the _general purpose_ language it must be a library concept, not a language concept. We might have a promise-like object in the Kotlin standard _library_ one day, but definitely not in the _language_.

It don't question your observation that Promise-based design has fewer primitives. The sound design is not about minimising the number of primitives. The sound design is about layering higher-level abstractions on top of an appropriate low-level abstractions. Kotlin coroutines provide a set of low-level abstractions that let you implement high-level Promise primitives on top of them without loosing efficiency. 

Minimisation of the number of primitives in the second revision of coroutines KEEP was not a goal. The goal was to make it closer to metal in terms of abstractions that it exposes and the reduction in complexity was just a side effect.

Having said that, we'd welcome any further reduction in complexity in the design, but introduction of any Promise-like construct into the language is out of the question and will not be considered.

**Comment by norswap** [comment]
> You can only reuse a promise object if you are absolutely sure that nobody is holding any references to the Promise object, because the holder of the reference can invoke then at any time.

True, and we seem to be in that very particular case here, so no worries. We just add an `internal` modifier in front of `proceed` in the `Yielder` class. (There is still the problem with its use from an inline function, but either a setter or `@PublishedApi` can resolve that.)

2. True again, at least in the general case. But if we have our own private subclass or Promise, which isn't leaked to other people, then we can safely break that rule. If you are thinking I'm riddling this thing with exceptions, hold on to that thought and seen my answer to point 3.

3. Not quite. This is the issue with describing a thing in a 15 minutes comment instead of a well thought-out proposal.

When I say, Promise is a primitive, I really mean the interface. Nothing really magical language-wise needs to happen in the implementation.

In fact, the implementations of `resolve`/`reject` correspond to the idea of dispatcher in the current design: you chose where to schedule the continuations. Running them immediately is a valid choice in that respect. (It does have the disadvantage you described, and that makes it inappropriate for certain use cases.)

I do agree it's not very nice to break the substitution principle (no multiple `then` for our local promise). There are ways to fix this (a class hierarchy of promises for instance). The key idea here is not the interface of Promise like in other language. That can change.

The key ideas is that of a trigger to continue some continuation. (1) you can add continuations (then, catch), (2) you can schedule them (resolve, reject). It makes sense that if only two actors share the trigger, then we can reuse it. Clearly, it's not an insurmountable problem: I'm pretty sure it can be done elegantly.

-----

My initial question was if I could submit a detailed proposal.

I appreciate you pointing out potential pitfalls. I know you did it to save me time and avoid me some frustration.

I think what came out of it is that if we can make lightweight/reusable promises (something of which you are doubtful), it can work.

I might go ahead with it, I'll think on the concepts and we shall see.

**Comment by elizarov** [comment]
It really worth the effort to write some nice and optimised Promise-based library for Kotlin coroutines. All the promise-like examples we have now are based on Java's `CompletableFuture` which is extremely heavy. I would really appreciate if you spend time to write some light-weight Promise library for Kotlin with the optimisations you have in mind. 

Promise is the right abstraction for many use-cases and having a light-weight Promise library with the built-in support of `async` and `await` primitives is a good idea. You can base your implementation on top of the low-level abstractions that Kotlin coroutines provide. If you study "Implementation details" section of KEEP you'll find there description of `CoroutineIntrinscis` object that gives you an unhindered access to the underlying machinery without any overhead whatsoever. If you base your `wait()` implementation directly on top of `CoroutineIntrinscis.suspendCoroutineOrReturn`, then you implementation has a potential to be way more efficient that the one based on a `CompletableFuture`.

**Comment by ilya-g** [comment]
@elizarov [kovenant](https://github.com/mplatvoet/kovenant) is an alternative for CompletableFutures
